### ISSUE TITLE
storage: MVCCGarbageCollect is slow in v22.2 compared to v22.1

### ISSUE THREAD

**Issue by aliher1911** [cockroachdb/cockroach#87277]
Running BenchmarkMVCCGarbageCollect (+modified) to include MVCCStats to make it more comparable to real runs:

```
22.1 (with stats update)

BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=2/deleteVersions=1-16            226299              4816 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=1-16         141714              7635 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=16-16         96882             11860 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=32-16         71068             16025 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=512-16         8449            142570 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=1015-16        3842            275308 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=1023-16        4149            274660 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=2/deleteVersions=1-16            746           1566034 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=1-16         212           5182729 ns/op
>BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=16-16         84          12559669 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=32-16         63          17568397 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=512-16         6         190913206 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=1015-16                        3         351730931 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=1023-16                        3         343130885 ns/op

22.1 (without stats update)

BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=2/deleteVersions=1-16            228370              4773 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=1-16         141380              7663 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=16-16        101109             11101 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=32-16         77886             14664 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=512-16         9637            125238 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=1015-16        4796            237608 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=1023-16        4208            242834 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=2/deleteVersions=1-16            774           1514281 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=1-16         220           5165355 ns/op
>BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=16-16         98          12588087 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=32-16         66          16583878 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=512-16         6         201204474 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=1015-16                        4         367584725 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=1023-16                        3         393604358 ns/op

Master (with stats)

BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=2/deleteVersions=1/numRangeTs=0/updateStats=true-16                273270          4243 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=1/numRangeTs=0/updateStats=true-16             157464          7689 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=16/numRangeTs=0/updateStats=true-16             96703         12978 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=32/numRangeTs=0/updateStats=true-16             65997         17460 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=512/numRangeTs=0/updateStats=true-16             6937        174425 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=1015/numRangeTs=0/updateStats=true-16            3046        336960 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=1023/numRangeTs=0/updateStats=true-16            3135        339966 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=2/deleteVersions=1/numRangeTs=0/updateStats=true-16                717       1647826 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=1/numRangeTs=0/updateStats=true-16              64      17664351 ns/op
>BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=16/numRangeTs=0/updateStats=true-16              5     207204312 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=32/numRangeTs=0/updateStats=true-16              3     397054718 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=512/numRangeTs=0/updateStats=true-16             1    7044993137 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=1015/numRangeTs=0/updateStats=true-16                     1    14448873505 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=1023/numRangeTs=0/updateStats=true-16                     3     403105760 ns/op

Master (without stats)

BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=2/deleteVersions=1/numRangeTs=0/updateStats=false-16               287023          4198 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=1/numRangeTs=0/updateStats=false-16            152658          7607 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=16/numRangeTs=0/updateStats=false-16           102878         11380 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=32/numRangeTs=0/updateStats=false-16            74265         16096 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=512/numRangeTs=0/updateStats=false-16            7830        151733 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=1015/numRangeTs=0/updateStats=false-16           3432        297439 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1/numVersions=1024/deleteVersions=1023/numRangeTs=0/updateStats=false-16           3717        287882 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=2/deleteVersions=1/numRangeTs=0/updateStats=false-16               703       1584928 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=1/numRangeTs=0/updateStats=false-16             67      17417538 ns/op
>BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=16/numRangeTs=0/updateStats=false-16             5     201719601 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=32/numRangeTs=0/updateStats=false-16             3     397685799 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=512/numRangeTs=0/updateStats=false-16            1    7004730468 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=1015/numRangeTs=0/updateStats=false-16                    1    14312205532 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=1023/numRangeTs=0/updateStats=false-16                    3     351458143 ns/op
```

If we compare highlighted (with >) case:
```
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=16-16         84          12559669 ns/op
BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=16/numRangeTs=0/updateStats=true-16              5     207204312 ns/op
```
we see a 16x degradation which looks scary.

Flame graph of reruning only this test:
(running 22.1 release branch with just enginepb.MVCCStats pointer added to GC invocation)
```
go test ./pkg/storage/. -bench=BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=16 -run=XXXXX --test.cpuprofile mvcc-gc-cpu-release-good.pprof -v
```
<img width="1145" alt="Screenshot 2022-09-01 at 16 14 04" src="https://user-images.githubusercontent.com/5124786/187953038-c6ed50c5-af31-4597-82d5-5df436d2dbe0.png">

And on master:
(with now extra options for range tombstones, but they are disabled)
```
go test ./pkg/storage/. -bench=BenchmarkMVCCGarbageCollect/pebble/keySize=128/valSize=128/numKeys=1024/numVersions=1024/deleteVersions=16/numRangeTs=0/updateStats=true -run=XXXXX --test.cpuprofile mvcc-gc-cpu-master-bad.pprof -v
```
<img width="1185" alt="Screenshot 2022-09-01 at 16 14 16" src="https://user-images.githubusercontent.com/5124786/187952931-1cd4472a-f9d8-4a2f-a52f-cf030f2fcd7e.png">

Jira issue: CRDB-19249

Epic CRDB-2624

**Comment by aliher1911** [comment]
For reference - this is a change to benchmark on master to add stats and range tombstone used to test #87261

**Comment by erikgrinaker** [comment]
Thanks for running these benchmarks! This seems fairly pathological, so we'll definitely need to address this for 22.2. Seems likely to be an issue in Pebble.

Maybe it'd be worthwhile to come up with a minimal repro benchmark?

Btw, for the future, we usually use [benchstat](https://pkg.go.dev/golang.org/x/perf/cmd/benchstat) to compare benchmark results, and do 5-10 runs of each benchmark to account for variance. It's also a good idea to stop any other processes on the system, and do another run first to warm up the OS/CPU for the workload -- I've seen this affect benchmarks as much as 10%. Something like:

```
$ make bench PKG=./pkg/storage BENCHES='BenchmarkMVCCGarbageCollect' TESTFLAGS="-v --count 10" | grep Benchmark >bench-22.1.txt
$ make bench PKG=./pkg/storage BENCHES='BenchmarkMVCCGarbageCollect' TESTFLAGS="-v --count 10" | grep Benchmark >bench-22.1.txt
$ git checkout master
$ make bench PKG=./pkg/storage BENCHES='BenchmarkMVCCGarbageCollect' TESTFLAGS="-v --count 10" | grep Benchmark >bench-master.txt
$ make bench PKG=./pkg/storage BENCHES='BenchmarkMVCCGarbageCollect' TESTFLAGS="-v --count 10" | grep Benchmark >bench-master.txt
$ benchstat bench-22.1.txt bench-master.txt
```

**Comment by aliher1911** [comment]
I tried a test with the same data that GC test has but that uses MVCCIterator and SeekLT for every object sequentially:
https://github.com/cockroachdb/cockroach/pull/87373

The results 22.1 vs master are:
name       old time/op  new time/op  delta
SeekLT-16   694ms ± 2%   836ms ± 3%  +20.46%  (p=0.000 n=10+10)
SeekGE-16   621ms ± 1%   757ms ± 1%  +21.95%  (p=0.000 n=10+9)

**Comment by aliher1911** [comment]
Those tests use inmem engine, maybe that makes difference more noticeable as we don't have any IO operations involved and impact of compares and such is more visible?

**Comment by erikgrinaker** [comment]
20% is significant and worth looking into. However, it's pretty far from 1600%, disk IO or not -- do you have any thoughts on where the discrepancy is coming from? Could we use the same engine setup for this benchmark and the GC benchmark to find the discrepancy? Also, enabling intents and range keys on the iterator might make a difference.

**Comment by aliher1911** [comment]
Looking on stacks, those are not the same paths that Next() takes. So it would be a red herring probably. As Jackson mentioned, Next does switchToMinHeap internally and that is calling batchskl.(*Iterator).Next which is causing slowness. It doesn't happen when I just do SeekGT/SeekLT. Naively iterating with Next() over the data set doesn't trigger this behaviour either.
I'll dig further to see what this is about.

**Comment by aliher1911** [comment]
I think the key to the slowness is updated batch. Tests with just seek or seeklt/next doesn't have significant slowdown by itself. But if test does:
```
 func(rw ReadWriter, iter MVCCIterator, key MVCCKey) error {
		iter.SeekLT(key)
		iter.Next()
		return rw.ClearMVCC(key)
})
```
for eligible keys, then degradation is significant:

name           old time/op  new time/op  delta
SeekLTNext-16   2.71s ± 2%  39.78s ± 4%  +1370.12%  (p=0.000 n=9+10)

**Comment by aliher1911** [comment]
To verify the idea that batch gives the worst effects:

```
name                          old time/op  new time/op  delta
SeekLTNext/useBatch=false-16   354ms ± 6%   397ms ± 4%    +12.16%  (p=0.000 n=10+10)
SeekLTNext/useBatch=true-16    2.58s ± 7%  38.74s ± 2%  +1402.47%  (p=0.000 n=10+10)
```

batch was also affecting performance in the 22.1 giving 7x slowdown, on master it is giving 98x slowdown.

**Comment by aliher1911** [comment]
cc: @jbowens

**Comment by erikgrinaker** [comment]
Had a closer look at the benchmark, and it's fairly trivial -- it uses a plain old `pebbleIterator` which essentially just calls straight through to the `pebble.Iterator`. Seems likely that the issue is with Pebble somehow, so I'm reassigning to storage for investigation.

**Comment by sumeerbhola** [comment]
@jbowens I see the multiple Pebble issues opened in response to this, but I don't quite understand what is it about this use case that is causing slowness for this release (compared to the past).

**Comment by jbowens** [comment]
@sumeerbhola — The change to batch mutation visibility (cockroachdb/pebble#1640) is new in this release. `MVCCGarbageCollect` exercises the pathological behavior described in cockroachdb/pebble#1944. We visit `GCKey`s in increasing order, so we're accumulating sorted point deletes to the batch. The initial `SeekGE` for each MVCC user key is fine—there's no skiplist entries with a greater key. But if we don't find the first garbage key through stepping, we `SeekLT` to find the first garbage key's predecessor. On the batch iterator, this `SeekLT` first finds the last written point tombstone and then needs to Prev through the entirety of the batch to find that there are no keys currently visible in the batch. We then `Next` to move onto the garbage key, which then requires scanning back through the entirety skiplist. Iterator bounds are enforced at each step.

I filed the issue cockroachdb/pebble#1640 as something we might want to explore in the longer term. For avoiding the regression this release, I'm going to look at whether we can enforce batch visibility at the merging iterator level rather than within the batch iterator itself. This would make batch visibility behave more like normal visibility. We don't need to step the batch iterator further unless we actually need to know if it contains a lessor key.

**Comment by jbowens** [comment]
Fixed by #87714.
